File: src/com/train/algorithm/search/implementInC++/CombinationSum.cpp
//
// Created by 罗斯 on 4/9/2022.
//
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> cur;
        sort(candidates.begin(),candidates.end());
        dfs(candidates,target,0,cur,res);
        return res;
    }
private:
    void dfs(vector<int>& candidates,int target,int s,vector<int>& cur,vector<vector<int>>& res){
        if(target == 0){
            res.push_back(cur);
            return;
        }

        for(int i = s;i < candidates.size();++i){
            if(candidates[i] > target) break;
            cur.push_back(candidates[i]);
            dfs(candidates,target - candidates[i],i,cur,res);
            cur.pop_back();
        }
    }
};

File: src/com/train/algorithm/search/implementInPython/CombinationSum.py
class Solution(object):

    def dfs(self, nums, target, path, res):
        if target < 0:
            return
        if target == 0:
            res.append(path)
            return
        for i in range(len(nums)):
            self.dfs(nums[i:],target - nums[i],path + [nums[i]],res)

    def combinationSum(self, candidates, target):
        res = []

        self.dfs(candidates,target,[],res)

        return res
